proposals/0000-fine-grained-undecidable.md

```diff
@@ -0,0 +1,185 @@
+
+## Proposed Change Specification
+
+We introduce new modifiers `%NoTerminationCheck`, `%LiberalCoverage`, `%LiberalInjectivity`, `%Overlapping`, `%Overlappable`, and `%Overlaps` which are used _before_ the keyword introducing the definition where we want to lift the corresponding restriction. The modifier syntax is taken from this [approved proposal](https://github.com/ghc-proposals/ghc-proposals/pull/370), with the additional [change](https://github.com/ghc-proposals/ghc-proposals/pull/392) of modifiers appearing _before_ the modified thing.
+
+1. Putting `%NoTerminationCheck` before a class declaration skips its superclass-cycle check.
+
+    ```haskell
+    %NoTerminationCheck ; class C (F a) => C a where ...
+    ```
+
+    ```diff
+    +cls_mod  : '%NoTerminationCheck'
+    +cls_mods : {- empty -} | cls_mods cls_mod
+
+    cl_decl :
+    - : 'class' tycl_hdr fds where_cls
+    + : cls_mods ';' 'class' tycl_hdr fds where_cls
+    ```
+
+2. Putting `%NoTerminationCheck` before the forall in a quantified constraint skips its termination check. 
+
+    ```haskell
+    f :: forall a. (%NoTerminationCheck forall b. C b a => C a a) => a -> a
+
+    -- Note the empty forall to have a target for the modifier
+    g :: forall a. (%NoTerminationCheck forall. D a a => C a) => a -> a
+    ```
+
+    ```diff
+    +forall_mod  : '%NoTerminationCheck'
+    +forall_mods : {- empty -} | forall_mods forall_mod
+
+    forall_telescope : 
+    - :             'forall' tv_bndrs '.'
+    + : forall_mods 'forall' tv_bndrs '.'
+    - |             'forall' tv_bndrs '->'
+    + | forall_mods 'forall' tv_bndrs '->'
+
+3. Putting `%NoTerminationCheck` before an instance declaration skips its termination check. Putting `%LiberalCoverage` allows for the more liberal coverage condition.
+
+    ```haskell
+    %NoTerminationCheck ; instance Eq (Tree a a) => Eq (Rose a) where ..
+    ```
+
+5. Putting `%NoTerminationCheck` within a type instance (or type in an associated type instance) skips the equation's termination check. Putting `%LiberalInjectivity` allows for the more liberal injectivity consistency check.
+
+    ```haskell
+    %NoTerminationCheck ; type instance F [a] = G a a
+
+    instance D (Maybe a) where
+      %NoTerminationCheck type F (Maybe a) = G a a
+    ```
+
+    The `%Overlapping`, `%Overlappable`, and `%Overlaps` modifiers keep the same meaning of the corresponding pragmas.
+
+    ```diff
+    +cls_inst_mod  : '%NoTerminationCheck' | '%LiberalCoverage'
+    +              | '%Overlapping' | '%Overlaps' | '%Overlappable'
+    +cls_inst_mods : {- empty -} | cls_inst_mods cls_inst_mod
+    +ty_fam_mod    : '%NoTerminationCheck' | '%LiberalInjectivity'
+    +ty_fam_mods   : {- empty -} | ty_fam_mods ty_fam_mod
+
+    inst_decl
+    - :               'instance' overlap_pragma inst_type where_inst
+    + : inst_mods ';' 'instance' overlap_pragma inst_type where_inst
```

Comment:
Should we not deprecate `{-# OVERLAPS #-}`  and `{-# OVERLAPPABLE #}` etc in favour of this new stuff?  The proposal should say.

Reply:
The proposal states:

> Note that the old pragmas are _not_ deprecated by this proposal.

although this is of course open to discussion.